### ISSUE TITLE
Basic URL Rewriting

### ISSUE THREAD

**Issue by rmetzler** [containous/traefik#133]
I'm struggling with doing the following with Traefik:

```
localhost/<docker-container-1>/ -> <docker-container-1>:<port>/
localhost/<docker-container-2>/ -> <docker-container-2>:<port>/
```

Then I came about [this issue](https://github.com/emilevauge/traefik/issues/73) where @emilevauge wrote the following:

> One more thing: for now, there is no URL rewriting mechanism in traefik, just to be clear ;)
> Your call to app.docker.com/userService will ends up at {container-ip}/userService.

I only had a glance on the source code but I think the statement is still true.

I'm not really sure though if this is consistent with the docs because I mistunderstand `PathPrefix` for the feature I'm looking for. https://github.com/emilevauge/traefik/blob/master/docs/index.md#-frontends

Could someone please clarify what exactly the difference between `Path` and `PathPrefix` rules is and  how I could reach my goal?

**Comment by emilevauge** [comment]
Hi @rmetzler, there is still no URL rewriting implementation in Traefik. What you want to do is not possible for now.
Here is the doc for [Path](http://www.gorillatoolkit.org/pkg/mux#Route.Path) and  [PathPrefix](http://www.gorillatoolkit.org/pkg/mux#Route.PathPrefix), Traefik uses Gorilla Mux as router.

**Comment by rmetzler** [comment]
I had a look into the source code and I think we would need to add some Req/Response rewriting middleware between frontend and backend in the following line: https://github.com/rmetzler/traefik/blob/master/traefik.go#L363

Do you have any preferences regarding functionality, flexibility, naming? I'll see what I'm able to do.

**Comment by emilevauge** [comment]
@rmetzler great, thanks :)
I don't have any preferences ;)
I suggest you to make a proposal in a comment in this issue, in order to validate the way to configure the rewriting.

**Comment by magnet** [comment]
Any news on this?

Here's how we currently nginx, to get an idea of what could be useful in URL rewriting in traefik: different micro-services own different top-level routes (/auth, /user, /query, etc). A micro-service instance might own multiple top-level routes. All service calls are prefixed with /service/ (e.g /service/auth, /service/user, /service/query). Everything except /services is re-routed to /index.html and handled by AngularJS's html5 router.

Having all services on the same domain is helpful with respect to browser security (cookies, CORS, etc) , allows HTTP connections from the browser to be reused (especially with http2), and generally works pretty well.

**Comment by bamarni** [comment]
@rmetzler : what you were looking for is now possible with `PathPrefixStrip` or `PathStrip` (see https://docs.traefik.io/basics/#frontends).

**Comment by dedalusj** [comment]
Not an expert on Traefik code but how about using a regexp for the rewrite?

A new frontend rule called `PathRewrite` can be introduced with syntax 
    
    PathRewrite: '/product/(.*)/view' '/item/$1/view'

This would result in the URL `/product/r14df/view` being mapped to `/item/r14df/view`.

**Comment by soerenmartius** [comment]
[Zalandos Skipper](https://github.com/zalando/skipper) implemented proxy as well as redirect methods. I think it would be a good idea to do something similar in this project since this is the only reason why we still stick to skipper.

**Comment by jvence** [comment]
+1

**Comment by ldez** [comment]
Closed by #1638, in the version 1.3.0-rc3.
